pkg/controller/handlers/slackreceiver/slackreceiver.go

```diff
@@ -47,3 +71,146 @@ func CreateOAuthApp(req router.Request, _ router.Response) error {
 
 	return nil
 }
+
+func (s *Handler) SubscribeToSlackEvents(req router.Request, resp router.Response) error {
+	slackReceiver := req.Object.(*v1.SlackReceiver)
+
+	if s.subscribed[slackReceiver.Name] != nil {
+		return nil
+	}
+
+	cred, err := s.gptScript.RevealCredential(req.Ctx, []string{system.OAuthAppPrefix + slackReceiver.Spec.ThreadName}, string(types.OAuthAppTypeSlack))
+	if err != nil {
+		return err
+	}
+
+	slackAppToken := cred.Env["APP_TOKEN"]
+
+	if slackAppToken == "" {
+		return nil
+	}
+
+	slackClient := slack.New("", slack.OptionAppLevelToken(slackAppToken))
+	client := socketmode.New(slackClient)
+
+	h := &eventHandler{
+		ctx:     req.Ctx,
+		storage: s.storage,
+	}
+
+	socketmodeHandler := socketmode.NewSocketmodeHandler(client)
+
+	socketmodeHandler.Handle(socketmode.EventTypeConnecting, middlewareConnecting)
+	socketmodeHandler.Handle(socketmode.EventTypeConnectionError, middlewareConnectionError)
+	socketmodeHandler.Handle(socketmode.EventTypeConnected, middlewareConnected)
+
+	// Handle a specific event from EventsAPI
+	socketmodeHandler.HandleEvents(slackevents.AppMention, h.middlewareAppMentionEvent)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s.subscribed[slackReceiver.Name] = cancel
+
+	go func() {
+		if err := socketmodeHandler.RunEventLoopContext(ctx); err != nil {
+			fmt.Printf("error running event loop: %v", err)
+		}
+		delete(s.subscribed, slackReceiver.Name)
+		resp.RetryAfter(time.Second * 10)
```

Comment:
I don't think this will work after all the controller return. There isn't anything looking at this anymore.

Reply:
I looked into this and this won't work like you expect. If you really need to look at this again, you'll have to do something else.

Reply:
Sorry missed this. I am going to update the status of the resource properly when connection fails so that it will reenqeue the object properly and retry in the goroutine. The goal is to provide a status error when connection to slack failed and controller retry properly.